src/Admin/Input/Form.php

```diff
@@ -0,0 +1,274 @@
+<?php
+/**
+ * Class Form
+ *
+ * @package Automattic\WooCommerce\Pinterest\Admin\Input
+ */
+
+declare( strict_types=1 );
+
+namespace Automattic\WooCommerce\Pinterest\Admin\Input;
+
+use Automattic\WooCommerce\Pinterest\Exception\ValidateInterface;
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Class Form
+ */
+class Form implements FormInterface {
+
+	use ValidateInterface;
+
+	/**
+	 * Form name.
+	 *
+	 * @var string
+	 */
+	protected $name = '';
+
+	/**
+	 * Form data.
+	 *
+	 * @var mixed
+	 */
+	protected $data;
+
+	/**
+	 * Children.
+	 *
+	 * @var FormInterface[]
+	 */
+	protected $children = array();
+
+	/**
+	 * Parent form.
+	 *
+	 * @var FormInterface
+	 */
+	protected $parent;
+
+	/**
+	 * Is form submitted.
+	 *
+	 * @var bool
+	 */
+	protected $is_submitted = false;
+
+	/**
+	 * Form constructor.
+	 *
+	 * @param mixed $data Form data.
+	 */
+	public function __construct( $data = null ) {
+		$this->set_data( $data );
+	}
+
+	/**
+	 * Get form name.
+	 *
+	 * @return string
+	 */
+	public function get_name(): string {
+		return $this->name;
+	}
+
+	/**
+	 * Set form name.
+	 *
+	 * @param string $name Form name.
+	 *
+	 * @return FormInterface
+	 */
+	public function set_name( string $name ): FormInterface {
+		$this->name = $name;
+
+		return $this;
+	}
+
+	/**
+	 * Get form children.
+	 *
+	 * @return FormInterface[]
+	 */
+	public function get_children(): array {
+		return $this->children;
+	}
+
+	/**
+	 * Add a child form.
+	 *
+	 * @param FormInterface $form Form.
+	 *
+	 * @return FormInterface
+	 *
+	 * @throws FormException If form is already submitted.
+	 */
+	public function add( FormInterface $form ): FormInterface {
+		if ( $this->is_submitted ) {
+			throw FormException::cannot_modify_submitted();
+		}
+
+		$this->children[ $form->get_name() ] = $form;
+		$form->set_parent( $this );
+
+		return $this;
+	}
+
+	/**
+	 * Remove a child with the given name from the form's children.
+	 *
+	 * @param string $name Form name.
+	 *
+	 * @return FormInterface
+	 *
+	 * @throws FormException If form is already submitted.
+	 */
+	public function remove( string $name ): FormInterface {
+		if ( $this->is_submitted ) {
+			throw FormException::cannot_modify_submitted();
+		}
+
+		if ( isset( $this->children[ $name ] ) ) {
```

Comment:
```suggestion
		if ( $this->has( $name ) ) {
```
We already have it so why not - just a nitpick, safe to ignore :)